### PR TITLE
Adding postgres eventonica lesson

### PR DESCRIPTION
**Pull request by lemon-yellow**
I'd appreciate any feedback on this lesson! I am not using an ORM because I'm not sure if the cohort will have encountered ORMs by the time they do this project. I am assuming that they have some familiarity with SQL and basic queries. I am also assuming that they have a way to execute queries on their machine (e.g., psql, PGAdmin).

The challenge questions require learning how to use Postgres constraints and using SQL Joins. Please let me know if the difficulty level is appropriate.

This addresses https://github.com/Techtonica/curriculum/issues/1159

### CODE REVIEW

**Comment by lalpert** [comment]
Thank you so much for working on this! I don't have time to actually review it, will leave that to others, but I can answer your question about what they should know already (these are the SQL lessons I made last year):
https://github.com/Techtonica/curriculum/blob/master/databases/data-modeling-1.md
https://github.com/Techtonica/curriculum/blob/master/databases/sql-1.md
https://github.com/Techtonica/curriculum/blob/master/databases/data-modeling-2.md
https://github.com/Techtonica/curriculum/blob/master/databases/sql-2.md

Looking at what I had there, they are expected to know basic joins involving 2 tables but "junction tables" where you need to join 3 tables were left as an extra challenge. (Now that I think about it, I'd love to have all apprentices understand those though. That could be a later change I guess)

Correct call on ORMs, that would be a future lesson once they're comfortable with raw SQL.

File: projects/eventonica/eventonica-part6-postgres.md [hunk 1]
```diff
@@ -1,9 +1,51 @@
...
-
+Previously, your data was stored in memory in Express, so your data would disappear if the application restarted. For production applications, data must be stored in a more permanent way. In this lesson, you will move your EventRecommender data to a Postgres database and connect to that database in your Express APIs.
 
 ### Instructions
 
+1. Ensure that you have Postgres installed on your machine - see instructions [here](../../databases/installing-postgresql.md).
+
+1. Create a new database named `eventonica`.
+
+1. In your `eventonica` database, create a table named `users` that contains the same fields as your `User` class in `eventRecommender.js`. Make the `id` column a [primary key](https://www.postgresql.org/docs/12/ddl-constraints.html#DDL-CONSTRAINTS-PRIMARY-KEYS) and use the datatype [serial](https://www.postgresql.org/docs/12/datatype-numeric.html#DATATYPE-SERIAL) to create an auto-incrementing integer id. Try running the following SQL insert multiple times to see how the `serial` type works: `INSERT INTO users (name) values ('jane');`. Your table should have automatically filled the `id` field for you!
```
Comment (vegetabill) [inline projects/eventonica/eventonica-part6-postgres.md]: Would you mind breaking up this longer step into sub-steps

If you have time, including a DDL snippet of the table might be nice too

File: projects/eventonica/eventonica-part6-postgres.md [hunk 2]
```diff
@@ -1,9 +1,51 @@
...
+
```
Comment (vegetabill) [inline projects/eventonica/eventonica-part6-postgres.md]: Would you mind breaking up this longer step into sub-steps

If you have time, including a DDL snippet of the table might be nice too

File: projects/eventonica/eventonica-part6-postgres.md [hunk 3]
```diff
@@ -1,9 +1,51 @@
...
+1. Create a table named `events` that contains the same fields as your `Event` class in `eventRecommender.js`. Create the `id` column like you did for the `users` table.
+
+1. Install [pg-promise](https://expressjs.com/en/guide/database-integration.html#postgresql) in your project folder - this module connects your Express application to a Postgres database.
+
+1. Copy the setup instructions for `pg-promise` in your `index.js` file. Your connection string is probably something like `postgres://{YOURNAME}@localhost:5432/eventonica`. If you used a password to protect your database, use `postgres://{YOURNAME}:{PASSWORD}@localhost:5432/eventonica`.
+
+1. Update your APIs to use SQL commands instead of the EventRecommender class. You can find example queries [here](https://github.com/vitaly-t/pg-promise/wiki/Learn-by-Example). Note that `pg-promise` requires you to specify how many rows, if any, a query should return. For example, `db.any` indicates that the query can return any number of rows, `db.one` indicates that the query should return a single row, and `db.none` indicates that the query must return nothing.
```
Comment (vegetabill) [inline projects/eventonica/eventonica-part6-postgres.md]: I like that this forces the apprentices to look it up. Just another style suggestion of breaking this into sub-steps.

File: projects/eventonica/eventonica-part6-postgres.md [hunk 4]
```diff
@@ -1,9 +1,51 @@
...
+    })
+    .catch(function(error) {
+        # on failure, send a generic 500 status
+        res.sendStatus(500)
+    });
+    ```
+
+1. Remove the EventRecommender import and instance from your `index.js` file.
+
+1. Test that your new APIs work using Postman and your webpage. Try restarting the Express application - your data should still be there!
```
Comment (vegetabill) [inline projects/eventonica/eventonica-part6-postgres.md]: Is it worth them checking using `psql` that the data is safely in postgres?

File: projects/eventonica/eventonica-part6-postgres.md (hunk 3 above)
Comment (vegetabill) [inline projects/eventonica/eventonica-part6-postgres.md]: I would suggest adding a note suggesting they write the commands in `psql` or a GUI SQL app before dropping them into JS. What do you think about that? How do you do this as a Real Engineer™?

File: projects/eventonica/eventonica-part6-postgres.md (hunk 2 above)
Comment (lemon-yellow) [reply projects/eventonica/eventonica-part6-postgres.md]: What do you mean by the DDL snippet?

File: projects/eventonica/eventonica-part6-postgres.md (hunk 4 above)
Comment (lemon-yellow) [reply projects/eventonica/eventonica-part6-postgres.md]: I don't think it's worth checking but I'll add a sentence to reify that the database is independent of the Express app.

File: projects/eventonica/eventonica-part6-postgres.md (hunk 3 above)
Comment (lemon-yellow) [reply projects/eventonica/eventonica-part6-postgres.md]: Makes sense to me - I will ask them to check `psql` or `PGAdmin`.

**Comment by lemon-yellow** [comment]
@vegetabill Thanks for your feedback! I made some changes based on your comments. I'm not certain what you mean by DDL snippet - can you elaborate?

File: projects/eventonica/eventonica-part6-postgres.md (hunk 2 above)
Comment (vegetabill) [reply projects/eventonica/eventonica-part6-postgres.md]: Sorry for using jargon unnecessarily. I believe it's the official name for the sql you can use to create the table, e.g.

```sql
CREATE TABLE SCOTT.DEPARTMENT
(DEPARTMENT_ID NUMBER,
   DNAME VARCHAR2 (100),
   LOCATION_ID NUMBER DEFAULT 100);
```

But now that I'm thinking about it, I guess it depends if we want that to be part of what they need to figure out on their own, so maybe ignore that suggestion!

**Pull request merged**